File: src/bundle_adjustment_inc.cpp
#include "opencv_all.hpp"
#include "cvsba.h"

int main(void)
{
    double camera_focal = 1000;
    cv::Point2d camera_center(320, 240);
    int n_views = 5;

    // Load multiple views of 'box.xyz'
    // c.f. You need to run 'image_formation.cpp' to generate point observation.
    //      You can apply Gaussian noise by change value of 'camera_noise' if necessay.
    std::vector<std::vector<cv::Point2d> > data;
    for (int i = 0; i < n_views; i++)
    {
        FILE* fin = fopen(cv::format("image_formation%d.xyz", i).c_str(), "rt");
        if (fin == NULL) return -1;
        std::vector<cv::Point2d> pts;
        while (!feof(fin))
        {
            double x, y, w;
            if (fscanf(fin, "%lf %lf %lf", &x, &y, &w) == 3)
                pts.push_back(cv::Point2d(x, y));
        }
        fclose(fin);
        data.push_back(pts);
        if (data.front().size() != data.back().size()) return -1;
    }

    // Assume that all cameras have the same and known camera matrix
    // Assume that all feature points are visible on all views
    cv::Mat K = (cv::Mat_<double>(3, 3) << camera_focal, 0, camera_center.x, 0, camera_focal, camera_center.y, 0, 0, 1);
    cv::Mat dist_coeff = cv::Mat::zeros(5, 1, CV_64F);
    std::vector<int> visible_all(data.front().size(), 1);

    // Prepare each camera projection matrix
    std::vector<cv::Mat> Ks, dist_coeffs, Rs, ts;
    std::vector<std::vector<int> > visibility;
    std::vector<std::vector<cv::Point2d> > xs;

    xs.push_back(data[0]);
    visibility.push_back(visible_all);
    Ks.push_back(K.clone());                                    // K for the first camera (index: 0)
    dist_coeffs.push_back(dist_coeff.clone());                  // dist_coeff for the first camera (index: 0)
    Rs.push_back(cv::Mat::eye(3, 3, CV_64F));                   // R for the first camera (index: 0)
    ts.push_back(cv::Mat::zeros(3, 1, CV_64F));                 // t for the first camera (index: 0)

    // 1) Select the best pair (skipped because all points are visible on all images)

    // 2) Estimate relative pose of the inital two views (epipolar geometry)
    cv::Mat F = cv::findFundamentalMat(data[0], data[1], cv::FM_8POINT);
    cv::Mat E = K.t() * F * K;
    cv::Mat R, t;
    cv::recoverPose(E, data[0], data[1], K, R, t);

    xs.push_back(data[1]);
    visibility.push_back(visible_all);
    Ks.push_back(K.clone());                                    // K for the second camera
    dist_coeffs.push_back(dist_coeff.clone());                  // dist_coeff for the second camera
    Rs.push_back(R.clone());                                    // R for the second camera
    ts.push_back(t.clone());                                    // t for the second camera

    // 3) Reconstruct 3D points of the initial two views (triangulation)
    cv::Mat Rt;
    cv::hconcat(R, t, Rt);
    cv::Mat P0 = K * cv::Mat::eye(3, 4, CV_64F);
    cv::Mat P1 = K * Rt, X;
    cv::triangulatePoints(P0, P1, xs[0], xs[1], X);

    std::vector<cv::Point3d> Xs;
    X.row(0) = X.row(0) / X.row(3);
    X.row(1) = X.row(1) / X.row(3);
    X.row(2) = X.row(2) / X.row(3);
    X.row(3) = 1;
    for (int c = 0; c < X.cols; c++)
        Xs.push_back(cv::Point3d(X.col(c).rowRange(0, 3)));

    // Incrementally add more views
    cvsba::Sba sba;
    cvsba::Sba::Params param;
    param.type = cvsba::Sba::MOTIONSTRUCTURE;
    param.fixedIntrinsics = 5;
    param.fixedDistortion = 5;
    param.verbose = true;
    sba.setParams(param);
    for (int i = 2; i < n_views; i++)
    {
        // 4) Select the next image to add (skipped because all points are visible on all images)

        // 5) Estimate relative pose of the next view (PnP)
        cv::Mat rvec;
        cv::solvePnP(Xs, data[i], K, dist_coeff, rvec, t);
        cv::Rodrigues(rvec, R);

        xs.push_back(data[i]);
        visibility.push_back(visible_all);
        Ks.push_back(K.clone());                                // K for the third and other cameras
        dist_coeffs.push_back(dist_coeff.clone());              // dist_coeff for the third and other cameras
        Rs.push_back(R.clone());                                // R for the third and other cameras
        ts.push_back(t.clone());                                // t for the third and other cameras

        // 6) Reconstruct newly observed 3D points (triangulation, skipped because all points are visible on all images)

        // 7) Optimize camera pose and 3D points (bundle adjustment)
        try { double error = sba.run(Xs, xs, visibility, Ks, Rs, ts, dist_coeffs); }
        catch (cv::Exception) { }
    }

    // Store the 3D points to an XYZ file
    FILE* fpts = fopen("bundle_adjustment_inc(point).xyz", "wt");
    if (fpts == NULL) return -1;
    for (size_t i = 0; i < Xs.size(); i++)
        fprintf(fpts, "%f %f %f\n", Xs[i].x, Xs[i].y, Xs[i].z);
    fclose(fpts);

    // Store the camera poses to an XYZ file 
    FILE* fcam = fopen("bundle_adjustment_inc(camera).xyz", "wt");
    if (fcam == NULL) return -1;
    for (size_t i = 0; i < Rs.size(); i++)
    {
        cv::Mat p = -Rs[i].t() * ts[i];
        fprintf(fcam, "%f %f %f\n", p.at<double>(0), p.at<double>(1), p.at<double>(2));
    }
    fclose(fcam);
    return 0;
}

File: examples/bundle_adjustment_inc.py
from scipy.optimize import least_squares
from scipy.sparse import lil_matrix
from scipy.spatial.transform import Rotation
import cv2
import numpy as np
import matplotlib.pyplot as plt
import time
import argparse
import open3d as o3d

input_num = 5
f, cx, cy = 1000, 320, 240

class BA():
    def __init__(self, f, cx, cy, n_cameras, n_points):
        self.K = np.array([[f, 0, cx], [0, f, cy], [0, 0, 1]])
        self.n_cameras = n_cameras
        self.n_points = n_points

    def sparsity_jacobian_inc(self, camera_indices, point_indices, n_scene):
        m = camera_indices.size * 2 # m = 800 x 2
        n = 6 * (n_scene + 1) + self.n_points * 3 # n = 6 * cameras = 5 + 3 * points = 160 --> 510
        A = lil_matrix((m, n), dtype=int)

        i = np.arange(camera_indices.size)
        for s in range(6):
            A[2 * i, camera_indices * 6 + s] = 1
            A[2 * i + 1, camera_indices * 6 + s] = 1

        for s in range(3):
            A[2 * i, n_scene * 6 + point_indices * 3 + s] = 1
            A[2 * i + 1, n_scene * 6 + point_indices * 3 + s] = 1

        return A

    def project(self, camera_params, points_3d):
        rvec, tvec = camera_params[0][:, :3], camera_params[0][:, 3:]  # tvec = 1x3 -> 160개로 늘려줘야 함.
        proj_points = Rotation.from_rotvec(rvec).apply(points_3d) # 160 x 3
        proj_points += tvec                                       
        proj_points = proj_points @ self.K.T
        proj_points /= proj_points[2, np.newaxis] # 애매하네....
        return proj_points[:, :2]

    @staticmethod
    def cost_func(params, points_2d, n_points, camera_indices, point_indices, n_scene, cls):
        """
        params = (camera_params + points_3d).ravel(). 
        n_cameras = cameras number, which is also index of scene
        cls is ba itself
        """
        points_2d = points_2d.reshape((n_points * (n_scene + 1) , 2))
        camera_params = params[: 6 * (n_scene + 1)].reshape((n_scene + 1, 6))
        points_3d = params[6 * (n_scene + 1): ].reshape((n_points, 3))
        proj_points = cls.project(camera_params[camera_indices], points_3d[point_indices])
        return (points_2d - proj_points).ravel()

def load_2d_points():
    global input_num
    """Load 2D points observed from multiple views"""
    xs = []
    for i in range(input_num):
        input = f"../bin/data/image_formation{i}.xyz"
        x = np.genfromtxt(input, delimiter=" ")
        x = x[:, :2]
        xs.append(x)

    return np.array(xs)


def main():
    global f, cx, cy
    parser = argparse.ArgumentParser(description="This is BA for incremental")
    VIEW_TEST = False
    VIEW_OPT = True

    points_2d = load_2d_points()
    n_camera = points_2d.shape[0]
    n_points = points_2d.shape[1]
    # Assumption
    # - All cameras have the same and known camera matrix
    # - All points are visible on all camera views

    # 1. Select best pair(skipped because all points are visible on all images)

    # 2. Estimate relative pose of the initial two views (epipolar geometry) 두 개에 대해서만 하네...
    F, _ = cv2.findFundamentalMat(points_2d[0], points_2d[1], cv2.FM_8POINT)
    K = np.array([[f, 0, cx], [0, f, cy], [0, 0, 1]], dtype=np.float32)
    E = K.T @ F @ K

    _, R, t, _ = cv2.recoverPose(E, points_2d[0], points_2d[1])
    rvec = Rotation.from_matrix(R).as_rotvec()
    cameras = np.zeros((n_camera, 6)) # rotation and translation
    cameras[1, :] = np.array([[rvec[0], rvec[1], rvec[2], t[0], t[1], t[2]]], dtype=object)

    # 3. Reconstruct 3D points of the initial two views (triangulation)
    P0 = K @ np.eye(3, 4, dtype=np.float32)
    Rt = np.hstack((R, t))
    P1 = K @ Rt
    X = cv2.triangulatePoints(P0, P1, points_2d[0].T, points_2d[1].T)
    X /= X[3]
    X = X.T

    points_3d = np.zeros((n_points, 3))
    for i in range(points_3d.shape[0]):
        points_3d[i, :] = X[i, :3]

    if VIEW_TEST:
        print("test")
        pcd = o3d.geometry.PointCloud()
        pcd.points = o3d.utility.Vector3dVector(points_3d)

    # Push constraints of two views
    ba = BA(f, cx, cy, n_camera, n_points)

    # 2개는 해줘야함.
    cam_indices = np.array([[]])
    for i in range(2):
        cam_indices = np.hstack((cam_indices, np.full((1, 160), i))) # 0x160, 1x160, ...

    point_indices = np.array([])
    for i in range(2):
        point_indices = np.hstack((point_indices, np.linspace(0, 159, 160, dtype=int))) # (0 ~ 159) * 5


    opts_points_3d = None
    # Incrementally add more views
    for j in range(2, n_camera):
        # 4. Select the next image to add (skipped because all points are visible on all images)
        
        # 5. Estimate relative pose of the next view (PnP)
        _, rvec, tvec = cv2.solvePnP(points_3d, points_2d[j], K, None)
        cameras[j, :, np.newaxis] = np.array([[rvec[0], rvec[1], rvec[2], tvec[0], tvec[1], tvec[2]]], dtype=object)
        # 6. Reconstruct newly observed 3D points (triangulation; skipped because all points are visible on all images)
        
        # 7. Optimize camera pose and 3D points together (BA)
        params = np.hstack((cameras[:j+1].ravel(), points_3d.ravel())) # j번째 이전까지 

        cam_indices = np.hstack((cam_indices, np.full((1, 160), j, dtype=int)))  # 
        point_indices = np.hstack((point_indices, np.linspace(0, 159, 160, dtype=int)))
        cam_indices = cam_indices.astype(int)
        point_indices = point_indices.astype(int)

        jac = ba.sparsity_jacobian_inc(camera_indices=cam_indices, point_indices=point_indices, n_scene=j)
        
        t = time.time()
        
        opt = least_squares(ba.cost_func, params, jac_sparsity=jac, method='trf', args=(points_2d[: j+1], n_points, cam_indices, point_indices, j, ba))
        
        print("calc time:", time.time() - t)
        
    # Store the 3D points to an XYZ file
    opt_cameras = opt.x[: 6 * n_camera].reshape((n_camera,6))
    opts_points_3d = opt.x[6 * n_camera :].reshape((n_points, 3))
    if VIEW_OPT:
        print("opt")
        pcd = o3d.geometry.PointCloud()
        pcd.points = o3d.utility.Vector3dVector(opts_points_3d)

    # Store the Camera poses to to an XYZ file
    # Store the 3D points to an XYZ file
    point_file = "../bin/data/bundle_adjustment_inc(point)_by_cjh.xyz"
    with open(point_file, 'wt') as f:
        for i in range(n_points):
            data = f"{opts_points_3d[i, 0]} {opts_points_3d[i, 1]} {opts_points_3d[i, 2]}\n"
            f.write(data)

    camera_file = "../bin/data/bundle_adjustment_inc(camera)_by_cjh.xyz"
    with open(camera_file, 'wt') as f:
        for i in range(n_camera):
            data = f"{opt_cameras[i, 0]} {opt_cameras[i, 1]} {opt_cameras[i, 2]} {opt_cameras[i, 3]} {opt_cameras[i, 4]} {opt_cameras[i, 5]}\n"
            f.write(data)

    o3d.visualization.draw_geometries([pcd])
    

if __name__ == "__main__":
    main()